tests/regression/test_fdm.py

```diff
@@ -0,0 +1,274 @@
+def mesh(request):
+    nx = 4
+    distribution = {"overlap_type": (DistributedMeshOverlapType.VERTEX, 1)}
+    m = UnitSquareMesh(nx, nx, quadrilateral=True, distribution_parameters=distribution)
+    if request.param == 3:
+        m = ExtrudedMesh(m, nx)
+
+    x = SpatialCoordinate(m)
+    xnew = as_vector([acos(1-2*xj)/pi for xj in x])
+    m.coordinates.interpolate(xnew)
+    return m
+
+
+@pytest.fixture
+def expected(mesh):
+    if mesh.topological_dimension() == 2:
+        return [5, 5, 5]
+    elif mesh.topological_dimension() == 3:
+        return [8, 8, 8]
+
+
+@pytest.mark.skipcomplex
+def test_p_independence(mesh, expected):
+    nits = []
+    for p in range(3, 6):
+        V = FunctionSpace(mesh, "Lagrange", p)
+        u = TrialFunction(V)
+        v = TestFunction(V)
+
+        a = inner(grad(v), grad(u))*dx
+        L = inner(v, Constant(1))*dx
+
+        asm = "firedrake.ASMStarPC"
+        subs = ("on_boundary",)
+        if mesh.topological_dimension() == 3:
+            asm = "firedrake.ASMExtrudedStarPC"
+            subs += ("top", "bottom")
+        bcs = [DirichletBC(V, zero(V.ufl_element().value_shape()), sub) for sub in subs]
+
+        uh = Function(V)
+        problem = LinearVariationalProblem(a, L, uh, bcs=bcs)
+        solver = LinearVariationalSolver(problem, solver_parameters={
+            "mat_type": "matfree",
+            "ksp_type": "cg",
+            "ksp_atol": 0.0E0,
+            "ksp_rtol": 1.0E-8,
+            "ksp_norm_type": "unpreconditioned",
+            "ksp_monitor_true_residual": None,
+            "ksp_converged_reason": None,
+            "pc_type": "python",
+            "pc_python_type": "firedrake.P1PC",
+            "pmg_mg_levels": {
+                "ksp_type": "chebyshev",
+                "esteig_ksp_type": "cg",
+                "esteig_ksp_norm_type": "unpreconditioned",
+                "ksp_chebyshev_esteig": "0.8,0.2,0.0,1.0",
+                "ksp_chebyshev_esteig_noisy": True,
+                "ksp_chebyshev_esteig_steps": 8,
+                "ksp_norm_type": "unpreconditioned",
+                "pc_type": "python",
+                "pc_python_type": "firedrake.FDMPC",
+                "fdm": {
+                    "ksp_type": "preonly",
+                    "pc_type": "python",
+                    "pc_python_type": asm,
+                    "pc_star_backend": "petscasm",
+                    "pc_star_sub_sub_pc_type": "cholesky",
+                    "pc_star_sub_sub_pc_mat_factor_type": "cholmod",
+                }
+            },
+            "pmg_mg_coarse": {
+                "mat_type": "aij",
+                "ksp_type": "preonly",
+                "pc_type": "python",
+                "pc_python_type": "firedrake.AssembledPC",
+                "assembled_pc_type": "cholesky",
+            }})
+        solver.solve()
+        nits.append(solver.snes.ksp.getIterationNumber())
+    assert (nits == expected)
+
+
+@pytest.mark.skipcomplex
```

Comment:
The problem in complex mode should have a real solution. I am not entirely sure but it might be that something was breaking in complex mode, either PMGPC or FDMPC.